### PR TITLE
[RISCV] Add +optimized-nfN-segment-load-store

### PR DESCRIPTION
**Pull request by lukel97**
This is a follow up to #111511, where after benchmarking we learnt that the Banana Pi F3 has fast segmented loads for not just NF=2, but also NF=3 and NF=4: https://github.com/preames/bp3-microarch#vlseg_lmul_x_sew_throughput

This adds tuning features to allow these segment loads and stores to be costed cheaper and enables it for the spacemit-x60.

It also enables +optimized-nf2-segment-load-store by default in the generic tuning to maintain the previous behaviour when compiled without -mcpu or -mtune.

### CODE REVIEW

**Comment by lukel97** [comment]
cc @camel-cdr

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 1]
```diff
@@ -1386,6 +1386,13 @@ def TuneOptimizedZeroStrideLoad
                       "true", "Optimized (perform fewer memory operations)"
                       "zero-stride vector load">;
 
+foreach nf = {3-4} in
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:1389]: Any reason not to support the option 3-8?  We'd only default x80 to it for 3/4, but would having the others be useful for testing?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -738,8 +738,11 @@ InstructionCost RISCVTTIImpl::getInterleavedMemoryOpCost(
                                             AddressSpace, DL)) {
 
         // Most available hardware today optimizes NF=2 as as one wide memory op
-        // + Factor * LMUL shuffle ops.
-        if (Factor == 2) {
+        // + Factor * LMUL shuffle ops. Some processors may also optimize NF=3
+        // and NF=4.
+        if (Factor == 2 ||
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:743]: Maybe extract a helper hasOptimizedSegmentLoadStore(unsigned Factor) which contains the switch?

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:1389]: No particular reason, happy to add those in. 

Another related thought: Should we also have a feature for NF=2 and enable it by default in the generic tuning? That way processors could opt out of it too.

File: llvm/test/Transforms/LoopVectorize/RISCV/interleaved-cost.ll [hunk 3]
```diff
@@ -1,33 +1,59 @@
 ; REQUIRES: asserts
-; RUN: opt -passes=loop-vectorize -mtriple=riscv64 -mattr=+v -debug-only=loop-vectorize -disable-output < %s 2>&1 | FileCheck %s --check-prefixes=CHECK,NO-OPT
-; RUN: opt -passes=loop-vectorize -mtriple=riscv64 -mattr=+v,+optimized-nf3-segment-load-store -debug-only=loop-vectorize -disable-output < %s 2>&1 | FileCheck %s --check-prefixes=CHECK,OPT-NF3
-; RUN: opt -passes=loop-vectorize -mtriple=riscv64 -mattr=+v,+optimized-nf4-segment-load-store -debug-only=loop-vectorize -disable-output < %s 2>&1 | FileCheck %s --check-prefixes=CHECK,OPT-NF4
+; RUN: opt -passes=loop-vectorize -mtriple=riscv64 -mattr=+v,-optimized-nf2-segment-load-store -debug-only=loop-vectorize -disable-output < %s 2>&1 | FileCheck %s --check-prefix=NO-OPT
+; RUN: opt -passes=loop-vectorize -mtriple=riscv64 -mattr=+v -debug-only=loop-vectorize -disable-output < %s 2>&1 | FileCheck %s --check-prefix=OPT-NF2
```
Comment (preames) [inline llvm/test/Transforms/LoopVectorize/RISCV/interleaved-cost.ll:3]: I think you swapped your runlines vs checks for this line and the one above.

Comment (lukel97) [reply llvm/test/Transforms/LoopVectorize/RISCV/interleaved-cost.ll:3]: +optimized-nf2-segment-load-store is enabled by default with the generic tuning, so the first run line is a negative feature to turn it off. The second run line actually has it enabled but I could make it explicit here if prefered

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 4]
```diff
@@ -716,6 +716,28 @@ RISCVTTIImpl::getMaskedMemoryOpCost(unsigned Opcode, Type *Src, Align Alignment,
   return getMemoryOpCost(Opcode, Src, Alignment, AddressSpace, CostKind);
 }
 
+static bool hasOptimizedSegmentLoadStore(unsigned NF,
+                                         const RISCVSubtarget *ST) {
+  switch (NF) {
+  case 2:
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:722]: Save some lines via using macros?

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 5]
```diff
@@ -1386,6 +1386,13 @@ def TuneOptimizedZeroStrideLoad
                       "true", "Optimized (perform fewer memory operations)"
                       "zero-stride vector load">;
 
+foreach nf = {2-8} in
+  def TuneOptimizedNF#nf#SegmentLoadStore :
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:1390]: A random thought: There are 250 features for RISCV now, we may need to increase `MAX_SUBTARGET_FEATURES` soon.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 4 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:722]: I considered it but remembered coming across some PRs where people were actively trying to remove uses of macros in LLVM, e.g. https://github.com/llvm/llvm-project/pull/105551

I'm not strongly opinionated on this, is there a general stance we should take here?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:722]: > is there a general stance we should take here?

Not found in https://llvm.org/docs/CodingStandards.html. I think it's OK not to use macros here, just a weak suggestion.

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 5 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:1390]: Isn't the limit 5*64 or 320?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:1390]: Yes, but that won't last long, especially if we add some fusions (like XiangShan).

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Review by preames — APPROVED** [review]
LGTM

**Comment by zqb-all** [comment]
LGTM

**Pull request merged**